api/server/sdk/volume_ops.go

```diff
@@ -0,0 +1,226 @@
+	if err == nil {
+		// Check the requested arguments match that of the existing volume
+		if spec.Size != v.GetSpec().GetSize() {
+			return nil, status.Errorf(
+				codes.AlreadyExists,
+				"Existing volume has a size of %v which differs from requested size of %v",
+				v.GetSpec().GetSize(),
+				spec.Size)
+		}
+		if v.GetSpec().GetShared() != req.GetSpec().GetShared() {
+			return nil, status.Errorf(
+				codes.AlreadyExists,
+				"Existing volume has shared=%v while request is asking for shared=%v",
+				v.GetSpec().GetShared(),
+				req.GetSpec().GetShared())
+		}
+		if v.GetSource().GetParent() != source.GetParent() {
+			return nil, status.Error(codes.AlreadyExists, "Existing volume has conflicting parent value")
+		}
+
+		// Return information on existing volume
+		return &api.VolumeCreateResponse{
+			Id: v.GetId(),
+		}, nil
+	}
+
+	// Check if the caller is asking to create a snapshot or for a new volume
+	var id string
+	if source != nil && len(source.GetParent()) != 0 {
+		// Get parent volume information
+		parent, err := util.VolumeFromName(s.driver, source.Parent)
+		if err != nil {
+			return nil, status.Errorf(
+				codes.InvalidArgument,
+				"unable to get parent volume information: %s",
+				err.Error())
+		}
+
+		// Create a snapshot from the parent
+		id, err = s.driver.Snapshot(parent.GetId(), false, &api.VolumeLocator{
+			Name: volName,
+		})
+		if err != nil {
+			return nil, status.Errorf(
+				codes.Internal,
+				"unable to create snapshot: %s\n",
+				err.Error())
+		}
+	} else {
+		// Create the volume
+		id, err = s.driver.Create(locator, source, spec)
+		if err != nil {
+			return nil, status.Errorf(
+				codes.Internal,
+				"Failed to create volume: %v",
+				err.Error())
+		}
+	}
+
+	return &api.VolumeCreateResponse{
+		Id: id,
+	}, nil
+}
+
+// CreateSimpleVolume provides a simple API to create a volume
+func (s *VolumeServer) CreateSimpleVolume(
+	ctx context.Context,
+	req *api.VolumeCreateSimpleVolumeRequest,
+) (*api.VolumeCreateResponse, error) {
+
+	// Generate objects from the parameters passed in
+	spec, locator, source, err := s.specHandler.SpecFromOpts(req.GetParameters())
+	if err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"Unable to get parameters: %s\n",
+			err.Error())
+	}
+	locator.Name = req.GetName()
+	spec.Size = uint64(req.GetSize())
+
+	// Create volume
+	return s.Create(ctx, &api.VolumeCreateRequest{
+		Locator: locator,
+		Source:  source,
+		Spec:    spec,
+	})
+}
+
+// Delete deletes a volume
+func (s *VolumeServer) Delete(
+	ctx context.Context,
+	req *api.VolumeDeleteRequest,
+) (*api.VolumeDeleteResponse, error) {
+
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Must supply volume id")
+	}
+
+	// If the volume is not found, return OK to be idempotent
+	volumes, err := s.driver.Inspect([]string{req.GetVolumeId()})
+	if (err == nil && len(volumes) == 0) ||
+		(err != nil && err == kvdb.ErrNotFound) {
```

Comment:
why check for kvdb.ErrNotFound? I don't think all drivers return that error. volume.ErrEnoEnt would be the correct error